### ISSUE TITLE
kube-proxy v1.26 blocks host services

### ISSUE THREAD

**Issue by stephan2012** [kubernetes/kubernetes#117613]
### What happened?

We run some Kubernetes “single-node clusters” that combine the control-plane and the workload onto a single host besides some services installed directly on the operating system (SSH, BIND9). With the upgrade to v1.26, all local services become unreachable as soon as kube-proxy configures the network stack. This happens because kube-proxy adds the node’s IP address to the ipset `KUBE-IPVS-IPS`.

We configure the node IP address explicitly for `kubelet`:

```shell
KUBELET_EXTRA_ARGS=--container-runtime-endpoint=unix:///run/cri-dockerd.sock --node-ip=10.169.8.6
```

Also, `kube-proxy` is configured to exclude the node IP address explicitly:

```yaml

    ipvs:
      excludeCIDRs:
      - 10.169.8.6/32
```

@liaralabs has collected more information in https://github.com/k3s-io/k3s/issues/7183#issuecomment-1491205024.

Overriding the `iptables` rules from `kube-proxy` with a higher-priority rules makes the services reachable again. Also, removing the node’s IP address from the `ipset` helps (until kube-proxy syncs the set again).

### What did you expect to happen?

kube-proxy should keep away from node-local services.

### How can we reproduce it (as minimally and precisely as possible)?

- Setup a minimal one-node cluster with `kubeadm` and configure it for `ipvs`.
- Add a `Service` of type `LoadBalancer` with `externalIPs` configured.

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.4", GitCommit:"f89670c3aa4059d6999cb42e23ccb4f0b9a03979", GitTreeState:"clean", BuildDate:"2023-04-12T12:13:53Z", GoVersion:"go1.19.8", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.4", GitCommit:"f89670c3aa4059d6999cb42e23ccb4f0b9a03979", GitTreeState:"clean", BuildDate:"2023-04-12T12:05:35Z", GoVersion:"go1.19.8", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
Azure, but likely affects any system.
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.2 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.2 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy

$ uname -a
Linux bbs 5.19.0-40-generic #41~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Mar 31 16:00:14 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Install tools

<details>
kubeadm
</details>

### Container runtime (CRI) and version (if applicable)

<details>
Docker + cri-docker
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
N/A
</details>

**Comment by stephan2012** [comment]
Downgrading to v1.25 helped to become the systems operational again.

**Comment by uablrek** [comment]
Most likely related https://github.com/kubernetes/kubernetes/issues/114815, please see especially https://github.com/kubernetes/kubernetes/issues/114815#issuecomment-1383804926. This PR https://github.com/kubernetes/kubernetes/pull/115019 could fix the problem in one-node systems.

**Comment by uablrek** [comment]
> kube-proxy should keep away from node-local services

What is "node-local services"? A service that has the (real) node address as externalIP or loadBalancerIP?

**Comment by stephan2012** [comment]
> What is "node-local services"? A service that has the (real) node address as externalIP or loadBalancerIP?

Let me rephrase: `kube-proxy` shall not interfere with applications running on the host. I.e., the node IP address should never be blocked even if used as `externalIP` or `loadBalancerIP`. Also, the IP should be completely ignored for the packetfilter when listed in `excludeCIDRs`.

**Comment by uablrek** [comment]
triage/accepted

**Comment by uablrek** [comment]
PR https://github.com/kubernetes/kubernetes/pull/115019 is updated and seem to work on K8s master.

**Comment by uablrek** [comment]
```
    ipvs:
      excludeCIDRs:
      - 10.169.8.6/32
```
This is [documented](https://kubernetes.io/docs/reference/config-api/kube-proxy-config.v1alpha1/) as:
```
excludeCIDRs is a list of CIDR's which the ipvs proxier should not touch when cleaning up ipvs services.
```
From what I can see it has no relevance in this problem.

**Comment by uablrek** [comment]
This issue is fixed by https://github.com/kubernetes/kubernetes/pull/115019, which is merged and will be released in v1.28.

I leave this issue open since I suspect there will be a request for back port. I don't know the procedure for that.

**Comment by thockin** [comment]
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-release/cherry-picks.md

**Comment by uablrek** [comment]
@stephan2012 A more or less unrelated topic in case you haven't noticed it: If you don't use nodePorts you can [turn off allocation](https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-nodeport-allocation) from v1.24.

**Comment by stephan2012** [comment]
> @stephan2012 A more or less unrelated topic in case you haven't noticed it: If you don't use nodePorts you can [turn off allocation](https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-nodeport-allocation) from v1.24.

@uablrek Thanks for the pointer, much appreciated.

**Comment by aojea** [comment]
> This issue is fixed by #115019, which is merged and will be released in v1.28.
> 
> I leave this issue open since I suspect there will be a request for back port. I don't know the procedure for that.

I think that we can close it now

/close
